scripts/tests/test_generate_changelog.py

```
@@ -0,0 +1,117 @@
+    def test_applies_explicit_release_date_to_target_heading(self) -> None:
+        """Metadata, rather than the process clock, owns the release date."""
+        changelog = "# Changelog\n\n## [1.0.0-rc1] - 2099-01-01\n\n- Fixed.\n"
+
+        updated = generate_changelog.apply_release_date(changelog, "1.0.0-rc1", date(2026, 7, 21))
+
+        self.assertIn("## [1.0.0-rc1] - 2026-07-21", updated)
+        self.assertNotIn("2099-01-01", updated)
+
+    def test_rejects_missing_target_heading(self) -> None:
+        """Malformed tool output cannot replace the prior artifact."""
+        with self.assertRaisesRegex(generate_changelog.ChangelogError, "found 0"):
```

Ruff (0.15.21)

[warning] Use `pytest.raises` instead of unittest-style `assertRaisesRegex`

Replace `assertRaisesRegex` with `pytest.raises`

(PT027)